pkg/integrations/manager.go

```diff
@@ -116,155 +113,238 @@ func (c *ManagerConfig) DefaultRelabelConfigs() ([]*relabel.Config, error) {
-	logger                log.Logger
-	integrations          map[Config]Integration
-	hostname              string
-	defaultRelabelConfigs []*relabel.Config
+	logger log.Logger
+
+	cfgMut sync.RWMutex
+	cfg    ManagerConfig
+
+	hostname string
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
 
 	im        instance.Manager
-	cancel    context.CancelFunc
-	done      chan bool
 	validator configstore.Validator
+
+	integrationsMut sync.RWMutex
+	integrations    map[string]*integrationProcess
 }
 
 // NewManager creates a new integrations manager. NewManager must be given an
 // InstanceManager which is responsible for accepting instance configs to
 // scrape and send metrics from running integrations.
 func NewManager(c ManagerConfig, logger log.Logger, im instance.Manager, validate configstore.Validator) (*Manager, error) {
-	integrations := make(map[Config]Integration)
-
-	for _, integrationCfg := range c.Integrations {
-		if integrationCfg.CommonConfig().Enabled {
-			l := log.With(logger, "integration", integrationCfg.Name())
-			i, err := integrationCfg.NewIntegration(l)
-			if err != nil {
-				return nil, err
-			}
-			integrations[integrationCfg] = i
-		}
-	}
+	ctx, cancel := context.WithCancel(context.Background())
 
-	return newManager(c, logger, im, integrations, validate)
-}
+	m := &Manager{
+		logger: logger,
+
+		ctx:    ctx,
+		cancel: cancel,
+
+		im:        im,
+		validator: validate,
+
+		integrations: make(map[string]*integrationProcess, len(c.Integrations)),
+	}
 
-func newManager(c ManagerConfig, logger log.Logger, im instance.Manager, integrations map[Config]Integration, validate configstore.Validator) (*Manager, error) {
-	defaultRelabels, err := c.DefaultRelabelConfigs()
+	var err error
+	m.hostname, err = instance.Hostname()
 	if err != nil {
-		return nil, fmt.Errorf("cannot get default relabel configs: %w", err)
+		return nil, err
 	}
 
-	ctx, cancel := context.WithCancel(context.Background())
+	if err := m.ApplyConfig(c); err != nil {
+		return nil, fmt.Errorf("failed applying config: %w", err)
+	}
+	return m, nil
+}
 
-	m := &Manager{
-		c:                     c,
-		logger:                logger,
-		integrations:          integrations,
-		defaultRelabelConfigs: defaultRelabels,
-		im:                    im,
-		cancel:                cancel,
-		done:                  make(chan bool),
-		validator:             validate,
+// ApplyConfig updates the configuration of the integrations subsystem.
+func (m *Manager) ApplyConfig(cfg ManagerConfig) error {
+	m.cfgMut.Lock()
+	defer m.cfgMut.Unlock()
+
+	m.integrationsMut.Lock()
+	defer m.integrationsMut.Unlock()
+
+	if cmp.Equal(m.cfg, cfg) {
+		return nil
+	}
+
+	select {
+	case <-m.ctx.Done():
+		return fmt.Errorf("Manager already stopped")
+	default:
+		// No-op
 	}
 
-	if c.UseHostnameLabel {
-		var err error
-		m.hostname, err = instance.Hostname()
+	// Iterate over our integrations. New or changed integrations will be
```

Comment:
I'm comfortable changing this into an implementation that builds up a new map and swaps them, though it'll still be very similar to how it works now; we don't want to stop all existing integrations to prevent any gaps in the metrics for integrations that didn't change.

Reply:
Great point